graf2d/gpadv7/src/TObjectDrawable.cxx

```diff
@@ -26,31 +29,124 @@
 
 using namespace ROOT::Experimental;
 
-std::string TObjectDrawable::DetectCssType(const std::shared_ptr<TObject> &obj)
+////////////////////////////////////////////////////////////////////
+/// Checks object ownership - used for TH1 directory handling and TF1 globals lists
+
+void TObjectDrawable::CheckOwnership(TObject *obj)
 {
-   const char *clname = obj ? obj->ClassName() : "TObject";
-   if (strncmp(clname, "TH1", 3) == 0) return "th1";
-   if (strncmp(clname, "TH2", 3) == 0) return "th2";
+   if (obj && obj->InheritsFrom("TH1")) {
+      TMethodCall call(obj->IsA(), "SetDirectory", "nullptr");
+      call.Execute((void *)(obj));
+   } else if (obj && obj->InheritsFrom("TF1")) {
+      TMethodCall call(obj->IsA(), "AddToGlobalList", "kFALSE");
+      call.Execute((void *)(obj));
+   }
+}
+
+////////////////////////////////////////////////////////////////////
+/// Provide css type
+
+std::string TObjectDrawable::DetectCssType(const TObject *obj)
+{
+   if (!obj) return "tobject";
+
+   const char *clname = obj->ClassName();
    if (strncmp(clname, "TH3", 3) == 0) return "th3";
+   if (strncmp(clname, "TH2", 3) == 0) return "th2";
+   if ((strncmp(clname, "TH1", 3) == 0) || obj->InheritsFrom("TH1")) return "th1";
    if (strncmp(clname, "TGraph", 6) == 0) return "tgraph";
    if (strcmp(clname, "TLine") == 0) return "tline";
    if (strcmp(clname, "TBox") == 0) return "tbox";
    return "tobject";
 }
 
+////////////////////////////////////////////////////////////////////
+/// Constructor, clones TObject instance
+
+TObjectDrawable::TObjectDrawable(TObject &obj) : RDrawable(DetectCssType(&obj))
+{
+   fKind = kObject;
+   auto clone = obj.Clone();
+   CheckOwnership(clone);
+   fObj = std::shared_ptr<TObject>(clone);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor, clones TObject instance
+
+TObjectDrawable::TObjectDrawable(TObject &obj, const std::string &opt) : TObjectDrawable(obj)
+{
+   SetOpt(opt);
+}
+
+////////////////////////////////////////////////////////////////////
+/// Constructor, takes ownership over the object
+
+TObjectDrawable::TObjectDrawable(TObject *obj) : RDrawable(DetectCssType(obj))
```

Comment:
It might be better to signal that with
```suggestion
TObjectDrawable::TObjectDrawable(unique_ptr<TObject> obj) : RDrawable(DetectCssType(obj))
```
which you can then convert to a `shared_ptr`. But as internally you use `shared_ptr` anyway, maybe expose that in the interface?
```suggestion
TObjectDrawable::TObjectDrawable(shared_pt<TObject> obj) : RDrawable(DetectCssType(obj))
```